src/dbnode/storage/index.go

```diff
@@ -2098,13 +2095,80 @@ func (i *nsIndex) CleanupExpiredFileSets(t time.Time) error {
 	return i.deleteFilesFn(filesets)
 }
 
-func (i *nsIndex) CleanupDuplicateFileSets() error {
+func (i *nsIndex) CleanupCorruptedFileSets() error {
+	/*
+	   Corrupted index filesets can be safely cleaned up if its not
+	   the latest volume index per index volume type/block start combo.
+
+	   We are guaranteed not to be actively writing to an index fileset once
+	   we're already writing to later volume indices.
+	*/
 	fsOpts := i.opts.CommitLogOptions().FilesystemOptions()
-	infoFiles := i.readIndexInfoFilesFn(
-		fsOpts.FilePathPrefix(),
-		i.nsMetadata.ID(),
-		fsOpts.InfoReaderBufferSize(),
+	infoFiles := i.readIndexInfoFilesFn(fs.ReadIndexInfoFilesOptions{
+		FilePathPrefix:   fsOpts.FilePathPrefix(),
+		Namespace:        i.nsMetadata.ID(),
+		ReaderBufferSize: fsOpts.InfoReaderBufferSize(),
+		IncludeCorrupted: true,
+	})
+
+	var (
+		toDelete                      []string
+		corrupted                     []fs.ReadIndexInfoFileResult
+		latestBlockStart              int64
+		latestVolumeIndexByVolumeType = make(map[idxpersist.IndexVolumeType]int)
 	)
+	// NB: Info files should be ordered by block start.
+	for _, file := range infoFiles {
+		// This intentionally skips the latest block start as that's the active block.
+		if file.Info.BlockStart > latestBlockStart {
+			for volType := range latestVolumeIndexByVolumeType {
+				latestVolumeIndex := latestVolumeIndexByVolumeType[volType]
+				// Reset latest vol idx.
+				latestVolumeIndexByVolumeType[volType] = 0
+
+				if len(corrupted) == 0 {
+					continue
+				}
+				// NB: We only remove stale index filesets to avoid deleting
+				// one we're actively writing to.
+				for _, cf := range corrupted {
+					if cf.ID.VolumeIndex < latestVolumeIndex {
+						toDelete = append(toDelete, cf.AbsoluteFilePaths...)
+					}
```

Comment:
Comment at the beginning of the method talks about removing all except latest volume per block start/volume type. Should we also check here that `cf`'s volume type matches `volType`, since `corrupted` array can contain any volume type?